File: src/main/java/guru/drinkit/lucene/SearchService.java
package guru.drinkit.lucene;

import guru.drinkit.domain.Recipe;
import org.apache.lucene.analysis.Analyzer;
import org.apache.lucene.analysis.ru.RussianAnalyzer;
import org.apache.lucene.document.Document;
import org.apache.lucene.document.Field;
import org.apache.lucene.document.IntField;
import org.apache.lucene.document.TextField;
import org.apache.lucene.index.DirectoryReader;
import org.apache.lucene.index.IndexWriter;
import org.apache.lucene.index.IndexWriterConfig;
import org.apache.lucene.queryparser.classic.MultiFieldQueryParser;
import org.apache.lucene.queryparser.classic.ParseException;
import org.apache.lucene.search.IndexSearcher;
import org.apache.lucene.search.Query;
import org.apache.lucene.search.TopDocs;
import org.apache.lucene.store.RAMDirectory;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.util.Arrays;

/**
 * @author pkolmykov
 */
@Service
public class SearchService {

    public static void main(String[] args) throws IOException, ParseException {
        Analyzer analyzer = new RussianAnalyzer();
        final RAMDirectory index = new RAMDirectory();
        final IndexWriterConfig config = new IndexWriterConfig(analyzer);

        final IndexWriter indexWriter = new IndexWriter(index, config);
        final Recipe recipe = new Recipe();
        recipe.setId(1);
        recipe.setName("водка со льдом");
        recipe.setDescription("desc");
        final Document doc = new Document();
        doc.add(new IntField("id", recipe.getId(), Field.Store.YES));
        doc.add(new TextField("name", recipe.getName(), Field.Store.NO));
        doc.add(new TextField("desc", recipe.getDescription(), Field.Store.NO));
        indexWriter.addDocument(doc);
        indexWriter.commit();
//        indexWriter.close();
//        indexWriter.close();
        final Document doc2 = new Document();
        doc2.add(new IntField("id", recipe.getId(), Field.Store.YES));
        doc2.add(new TextField("name", "вискарь", Field.Store.NO));
        doc2.add(new TextField("desc", "вкусный вискарь", Field.Store.NO));
        indexWriter.addDocument(doc2);
        indexWriter.commit();
        final Query query = new MultiFieldQueryParser(new String[]{"name", "desc"}, analyzer).parse("вискарь водка");
        System.out.println(query);
        final IndexSearcher searcher = new IndexSearcher(DirectoryReader.open(index));
        final TopDocs res = searcher.search(query, 10);
        System.out.println(Arrays.toString(res.scoreDocs));
        System.out.println(searcher.doc(res.scoreDocs[0].doc));
    }

    public void indexRecipes(Iterable<Recipe> recipes) {
//        recipes.
//        transformToDocument();
    }
}

File: src/main/java/guru/drinkit/lucene/SearchService.kt
package guru.drinkit.lucene

import guru.drinkit.domain.Ingredient
import guru.drinkit.domain.Recipe
import guru.drinkit.repository.IngredientRepository
import guru.drinkit.repository.RecipeRepository
import org.apache.lucene.analysis.Analyzer
import org.apache.lucene.document.Document
import org.apache.lucene.document.Field
import org.apache.lucene.document.Field.Store.NO
import org.apache.lucene.document.StoredField
import org.apache.lucene.document.TextField
import org.apache.lucene.index.DirectoryReader
import org.apache.lucene.index.IndexWriter
import org.apache.lucene.queryparser.classic.MultiFieldQueryParser
import org.apache.lucene.search.IndexSearcher
import org.apache.lucene.store.Directory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import javax.annotation.PostConstruct

/**
 * @author pkolmykov
 */
@Service
open class SearchService @Autowired constructor(
        val recipeIndexWriter: IndexWriter,
        val analyzer: Analyzer,
        val recipeDirectory: Directory,
        val recipeRepository: RecipeRepository,
        val ingredientRepository: IngredientRepository
) {

    private lateinit var recipeSearcher: IndexSearcher

    @PostConstruct
    open fun index() {
        try {
            recipeIndexWriter.deleteAll()
            recipeIndexWriter.addDocuments(convertRecipes())
            recipeIndexWriter.commit()
            recipeSearcher = IndexSearcher(DirectoryReader.open(recipeDirectory))
        } catch (e: Exception) {
            e.printStackTrace()
            //todo rollbar
        }
    }

    private fun convertRecipes(): List<Document> {
        val recipes = recipeRepository.findAll()
        val ingredients = ingredientRepository.findAll().associateBy { it.id!! }
        return recipes.map { transformToDocument(it, ingredients) }
    }

    private fun transformToDocument(recipe: Recipe, ingredients: Map<Int, Ingredient>): Document {
        val document = Document()
        document.add(StoredField("id", recipe.id!!))
        recipe.name?.let { document.add(RecipeFields.localizedName.toField(it)) }
        recipe.originalName?.let { document.add(RecipeFields.originalName.toField(it)) }
        recipe.description?.let { document.add(RecipeFields.description.toField(it)) }
        recipe.ingredientsWithQuantities.asSequence()
                .map { it.ingredientId }
                .mapNotNull { ingredients[it] }
                .forEach { ingredient ->
                    ingredient.name?.let { document.add(RecipeFields.ingredientName.toField(it)) }
                    ingredient.alias?.forEach { document.add(RecipeFields.ingredientAlias.toField(it)) }
                }
        return document
    }

    private val recipeQueryParser = recipeQueryParser()

    fun findRecipes(searchString: String): Collection<Int> {
        val topDocs = recipeSearcher.search(recipeQueryParser.parse(searchString), 50)
        return topDocs.scoreDocs.map { recipeSearcher.doc(it.doc).getField("id").numericValue().toInt() }
    }

    fun indexRecipe(recipe: Recipe) {
        val ingredients = ingredientRepository.findAll().associateBy { it.id!! }
        recipeIndexWriter.addDocument(transformToDocument(recipe, ingredients))
        recipeIndexWriter.commit()
    }

    private fun recipeQueryParser(): MultiFieldQueryParser = MultiFieldQueryParser(
            RecipeFields.values().map { it.name }.toTypedArray(),
            analyzer,
            RecipeFields.values().associate { it.name to it.boost }
    )

    companion object {
        enum class RecipeFields(val boost: Float) {
            localizedName(2.9f),
            originalName(2.1f),
            description(0.1f),
            ingredientName(1.5f),
            ingredientAlias(1f);

            fun toField(value: String): Field {
                return TextField(this.name, value, NO)
            }
        }
    }
}
